mi/dataset/parser/wavss_a_dcl.py

```diff
@@ -289,39 +277,47 @@ def _build_parsed_values(self):
         """
         particle_parameters = super(WavssADclMeanDirectionalDataParticle, self)._build_parsed_values()
 
+        band_len = len(self.band_parameter_types)
+        if len(self.payload) < (band_len + 2):
+            raise RecoverableSampleException('missing bands particle map header data')
+
+        bands_header = self.payload[:band_len]
+        num_bands = int(self.payload[0])
+
+        expected_payload_len = band_len + num_bands * 3 + 2
+        if len(self.payload) != expected_payload_len:
+            raise RecoverableSampleException('unexpected number of mean-directional parameters (got %d, expected %d)' %
+                                             (len(self.payload), expected_payload_len))
+
         # append the band description parameters
-        for name, group, function in BANDS_PARTICLE_MAP:
-            particle_parameters.append(self._encode_value(name, self.raw_data.group(group), function))
+        for value, (name, ptype) in zip(bands_header, self.band_parameter_types):
+            particle_parameters.append(self._encode_value(name, value, ptype))
 
         # append the mean directional specific parameters
-        for name, group, function in MEAN_DIRECTION_PARTICLE_MAP:
-            particle_parameters.append(self._encode_value(name, self.raw_data.group(group), function))
+        mean_header = self.payload[band_len:]
 
-        number_bands = int(self.raw_data.group(NUMBER_COUNT_GROUP))
-
-        # split the array of floats
-        data_str = self.raw_data.group(0)[self.raw_data.end(SPREAD_DIR_GROUP) + 1:
-                                          self.raw_data.end(END_MEAN_DIR_ARRAY_GROUP)]
-        flt_array = WavssADclCommonDataParticle.string_to_float_array(data_str)
+        for value, (name, ptype) in zip(mean_header, self.parameter_types):
+            particle_parameters.append(self._encode_value(name, value, ptype))
 
         # split up the array into 3 arrays each number of bands in length, taking each 3rd item, size of array
         # checked in wavss parser
-        psd = flt_array[0:number_bands*3:3]
-        mean_dir = flt_array[1:number_bands*3:3]
-        dir_spread = flt_array[2:number_bands*3:3]
+        spectra_payload = self.payload[band_len + 2:]
+        psd = spectra_payload[0:num_bands * 3:3]
+        mean_dir = spectra_payload[1:num_bands * 3:3]
+        dir_spread = spectra_payload[2:num_bands * 3:3]
 
         # to match with non-directional data, the mean directional arrays must be padded with NaNs so they are
         # the same size
-        for i in range(number_bands, MEAN_DIR_NUMBER_BANDS):
+        for i in range(num_bands, MEAN_DIR_NUMBER_BANDS):
```

Comment:
I'm never sure if I really want to comment on this.

In python 2, "xrange" is preferable here as it will create an iterator rather than generating the complete list ahead of time.

In python 3, range returns an iterable like xrange. So, changing this to xrange makes this code not python 3 compatible but slightly more efficient in python 2...

If we think we're going to want to move to python 3 some day, we should stick to range or use something like this:

`from past.builtins import xrange
for i in xrange(10**8):`